docs/enhancements/0003-module-nmc-responsibilty-refactoring.md

```diff
@@ -0,0 +1,110 @@
+4. If there is a suitable KernelMapping, Version field is present in the Module, module loader version label is on the node and
+   its value is equal to the Version - update the spec
+5. If there is a suitable KernelMapping, Version field is present in the Module, module loader version label is on the node and 
+   its value is not equal to Module's version (meaning old version) - do nothing
+6. If there is a suitable KernelMapping, Version field is present in the Module, is missing on the node (meaning kernel module should
+   not be running on the node)  - delete the spec
+
+In this implementation, Module-NMC does not need to delete the spec, but in the 2 following cases:
+1. during ordered upgrade (see point 6 above)
+2. Module is deleted, and so the kernel module should be unloaded
+
+```mermaid
+flowchart TD
+Module[KMM Module]-->|Reconcile| MNC[Module-NMC controller]
+MNC-->|get nodes based on node selector| J1((.))
+J1-->|no KernelMapping for node's kernel| Done[Done]
+J1-->|found KernelMapping for node's kernel| J2((.))
+    J2-->|Version missing in Module| US[Update NMC Spec]
+    J2-->|Version present in Module| J3((.))
+        J3-->|module loader version label equals Version| US
+        J3-->|module loader version label not equal Version| Done
+        J3-->|module loader version label missing| DS[Delete NMC Spec]
+```
+
+
+### NMC controller decision-making flow
+
+NMC takes into account the NMC spec, status, node's status and node's ready timestamp to make decision whether to run worker pods, and whether to run unload or load
+worker pod
+
+1. If Node is not Ready/Schedulable - do nothing
+2. If NMC's status is missing and Node's kernel version equal to NMC's spec kernel version - run worker load pod
+3. If NMC's spec is missing, NMC's status is present and NMC's status kernel version equal to Node's kernel version - run worker unload pod
+4. If NMC's spec is present and NMC's status is present, and NMC spec differ from NMC status:
+   - if status kernel version equal to node's kernel version - run worker unload pod
+   - if spec's kernel version equal to node's kernel version - run worker load pod
+5. If NMC's spec is present and NMC's status is present, and NMC spec equal to NMC status and status timestamp older then node's Ready timestamp - run worker load pod
+
+```mermaid
+flowchart TD
+NMC[NodeModuleConfig]-->|Reconcile| NMCC[NCM controller]
+NMCC-->| get NMC's node| J1((.))
+J1-->|node is not Ready/Schedulable| Done[Done]
+J1-->|node is Ready/Schedulable| J2((.)) 
+J2-->|status missing| J3((.))
+   J3-->|node's kernel equals spec' kernel| WLP[Create Worker Load Pod]
+   J3-->|node's kernel differs spec' kernel| Done
+J2-->|spec missing| J4((.))
+   J4-->|node's kernel equals status' kernel| WUP[Create Worker UnLload Pod]
+   J4-->|node's kernel differs status' kernel| Done
+J2-->|spec and status differ| J4((.))
+   J5-->|status kernel equals node's kernel| WUP
+   J5-->|status kernel differs node's kernel| J6((.))
```

Comment:
How do we reach the J5 state? Looks like you removed some line and forgot to update its children.

Reply:
Fixed